plugin/template/README.md

```diff
@@ -0,0 +1,135 @@
+# template
+
+*template* allows for dynamic responses base on the incoming query.
+
+## Syntax
+
+~~~
+template CLASS TYPE [REGEX...] {
+    [answer section]
+    [answer section]
+    [additional section]
+    [...]
+    [rcode  responsecode]
+}
+~~~
+
+* **CLASS** the query class (usually IN or ANY)
+* **TYPE** the query type (A, PTR, ...)
+* **REGEX** regex that are matched against the incoming query. Specifying no regex matches everything.
+* `section` A RFC 1035 style fragment build by a go template that contains the answer.
+* `resplonsecode` A response code (NXDOMAIN, SERVFAIL, ...)
+
+At least one answer section or rcode is needed.
+
+## Templates
+
+Each answer section is a full-featured go templated with the following predefined data
+* `.Name` the query name, as a string
+* `.Class` the query class (usually `IN`)
+* `.Type` the RR type requested (e.g. `PTR`)
+* `.Match` an array of all matches. `index .Match 0` refers to the whole match.
+* `.Group` a map fo the named capture groups.
+* `.Message` the incoming DNS query message.
+* `.Question` the matched question section.
+
+The output of the template must be a RFC 1035 style RR line (commonly refered to as a "zone file").
+
+**WARNING** there is a syntactical problem with go templates and caddy config files. Expressions like `{{$var}}` will be interpreted as a reference to an envorionment variable while `{{ $var }}` will work. Try to avoid template variables.
+
+## Examples
+
+### Resolve .invalid as NXDOMAIN
+
+The `.invalid` domain is a reserved TLD to indicate non-existing domains.
+
+~~~ corefile
+. {
+    proxy . 8.8.8.8
+
+    template ANY ANY "[.]invalid[.]$" {
+      rcode NXDOMAIN
+      answer "invalid. 60 {{ .Class }} SOA a.invalid. b.invalid. (1 60 60 60 60)"
+    }
+}
+~~~
+
+1. A query to .invalid will result in NXDOMAIN (rcode)
+2. A dummy SOA record is send to hand out a TTL of 60s for caching
+3. Querying `.invalid` of `CH` will also cause a NXDOMAIN/SOA response
+
+### Block invalid search domain completions
+
+Imagine you run `example.com` with a datacenter `dc1.example.com`. The datacenter domain
+is part of the DNS search domain.
+However `something.example.com.dc1.example.com` would usually indicate a wrong autocomplete.
```

Comment:
autocomplete -> indicates a non fully qualified domain that inadvertely has the default domain added or search path added.